doc/adr/0012-commit-to-datatypes-in-cardano-api.rst

```diff
@@ -0,0 +1,141 @@
+.. _commit-to-datatypes-in-cardano-api:
+
+ADR 12: Commit to datatypes in cardano-api
+==========================================
+
+Date: 2022-10-03
+
+Author(s)
+---------
+
+koslambrou <konstantinos.lambrou@iohk.io>
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+Since the genesis of the ``plutus-apps`` repository, the components have been
+historically using the types in the ``plutus-ledger-api`` package (which is now
+part of ``plutus`` repository) in the off-chain part of a Plutus application.
+
+This was desirable in order to start designing a way to build Plutus application before the ``cardano-ledger`` actually supported the Alonzo-era features.
+Of course, this resulted in the unintented consequence that we used ``TxInfo`` types (types that are designed to be used in Plutus scripts) in off-chain code.
+This wouldn't has been a problem if there was 1:1 relationship between on-chain and off-chain types.
+However, that presumption is wrong.
+
+Let's take the example of the ``TxOut`` representation of ``plutus-ledger-api`` for ``PlutusV2``.
+
+.. code-block:: haskell
+
+  data TxOut = TxOut {
+    ...
+    txOutReferenceScript :: Maybe ScriptHash
+    }
+
+As we can see, the ``TxOut`` can optionnaly store the ``ScriptHash`` of the referenced script.
+However, that is *not* the adequate representation of a ``TxOut`` in a
+transaction given the ``cardano-ledger`` specification.
+The off-chain ``TxOut`` should instead be:
+
+.. code-block:: haskell
+
+  data TxOut = TxOut {
+    ...
+    txOutReferenceScript :: Maybe Script
+    }
+
+where the reference script field can store that *actual* script, not just the hash.
+
+This proved that we need to start moving away from `plutus-ledger-api` types in
+the off-chain part of Plutus applications, especially in components like the
+emulator and the chain indexer.
+
+Decision
+--------
+
+* We will remove our datatype representation of a Cardano transaction
+  (``Ledger.Tx.Internal.Tx`` in ``plutus-ledger``) and fully commit to
+  ``Cardano.Api.Tx.Tx era`` in the codebase.
+
+* We will replace any use of ``plutus-ledger-api`` types by ``cardano-api`` types
```

Comment:
Maybe should we move this after the introduction of `cardano-api-extended` and talk about "`cardano-api-extended` types" here, as we may encounter scenarios where we need a slightly different representation.